File: array/single-element-in-a-sorted-array.rb
#!/usr/bin/env ruby
#
# https://leetcode.com/problems/single-element-in-a-sorted-array/#/description
#

def single_non_duplicate nums
  start = 0
  _end = nums.length - 1
  while start < _end
    tmp_mid = (start + _end) / 2
    mid = tmp_mid + tmp_mid % 2
    if nums[mid] == nums[mid - 1]
      # left shifta
      _end = mid - 2
    elsif nums[mid] == nums[mid + 1]
      # right shift
      start = mid + 2
    else
      return nums[mid]
    end
  end

  return nums[start]
end

a = [0,1,1]
puts single_non_duplicate a

File: single-element-in-a-sorted-array.go
package main

import "fmt"

func main() {
	input := []int{3, 3, 7, 7, 10, 11, 11}
	fmt.Println(singleNonDuplicate(input))
}

func singleNonDuplicate(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}

	left := 0
	right := len(nums) - 1
	mid := left + (right-left)/2
	for left < right {
		mid = left + (right-left)/2
		if mid%2 == 1 {
			mid--
		}

		if nums[mid] != nums[mid+1] {
			right = mid
		} else {
			left = mid + 2
		}
	}

	return nums[left]
}
